### PR TITLE
Implement a script-plugin to use Detekt for Kotlin code analsysis

### PR DESCRIPTION
**Pull request by yevhenii-nadtochii**
This changeset implements `detekt-code-analysis` script-plugin. This plugin adds `detekt` Gradle task that is bound to `check` task. The task runs `detekt` analyzer upon Kotlin sources.

In this PR, only the most basic configuration of Detekt is performed. It excludes `*Test.kt` and `*Spec.kt` files for some rules.

## Breaking change

Presence of Detekt on a classpath of `buildSrc` breaks fully-qualified accessors to our dependency-objects from `plugins` section. Thus, plugins can only be applied by string ID. See [issue](https://github.com/gradle/gradle/issues/22532).

For example, the original `plugins` section from `base` repository:

```kotlin
plugins {
    // ...
    with(io.spine.internal.dependency.Protobuf.GradlePlugin) { id(id) }
    with(io.spine.internal.dependency.ErrorProne.GradlePlugin) { id(id) }
}
```

Becomes the following:

```kotlin
plugins {
    // ...
    id("com.google.protobuf") // io.spine.internal.dependency.Protobuf.GradlePlugin
    id("net.ltgt.errorprone") // io.spine.internal.dependency.ErrorProne.GradlePlugin
}
```

### CODE REVIEW

**Comment by yevhenii-nadtochii** [comment]
@alexander-yevsyukov Please take a look.

File: buildSrc/src/main/kotlin/detekt-code-analysis.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * This script-plugin sets up Kotlin code analyzing with Detekt.
+ *
+ * After apply, Detekt is configured to use `${rootDir}/config/quality/detekt-config.yml` file.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/detekt-code-analysis.gradle.kts:30]: "Applying" or "application".

File: buildSrc/src/main/kotlin/detekt-code-analysis.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * An example of appending a custom config file to the default one:
+ *
+ * ```
+ * detekt {
+ *     config.from("config/detekt-custom-config.yml")
+ * }
+ * ```
+ *
+ * In order to totally substite it, just overwrite the corresponding property:
```
Comment (armiol) [inline buildSrc/src/main/kotlin/detekt-code-analysis.gradle.kts:42]: "substitute". 

Please have another proofread run.

**Review by armiol — CHANGES_REQUESTED** [review]

**Comment by yevhenii-nadtochii** [comment]
@armiol @alexander-yevsyukov PTAL.

I've updated the PR description to include a finding from this [prototype](https://github.com/SpineEventEngine/base/pull/756).

I don't see other options to set up Detekt from `buildSrc` without its presence on the classpath. Groovy Interop can't be used to apply a published plugin with a version from a script-plugin. I just remembered a similar [situation](https://github.com/SpineEventEngine/config/issues/298), when it wasn't possible to obtain a typed plugin on the classpath. But in that case, Grovvy Interop managed to solve the problem, as the requested plugin was meant to be applied in advance.

**Review by armiol — APPROVED** [review]

**Pull request merged**